stl/inc/xutility

```diff
@@ -4815,135 +4836,109 @@ namespace ranges {
-    // return lex_compare optimization category for contiguous iterators and ranges::less
-    return _Lex_compare_check_element_types<less<int>, remove_reference_t<_Iter_ref_t<_CtgIt1>>,
-        remove_reference_t<_Iter_ref_t<_CtgIt2>>, void>{};
-}
+template <class _Elem1, class _Elem2>
+struct _Lex_compare_memcmp_classify_pred<_Elem1, _Elem2, _RANGES less> {
+    using _Pred = conditional_t<_Lex_compare_memcmp_classify_elements<_Elem1, _Elem2>, less<int>, void>;
+};
 
-template <class _CtgIt1, class _CtgIt2, enable_if_t<_Iterators_are_contiguous<_CtgIt1, _CtgIt2>, int> = 0>
-constexpr auto _Lex_compare_memcmp_classify(const _CtgIt1&, const _CtgIt2&, const _RANGES greater&) {
-    // return lex_compare optimization category for contiguous iterators and ranges::greater
-    return _Lex_compare_check_element_types<greater<int>, remove_reference_t<_Iter_ref_t<_CtgIt1>>,
-        remove_reference_t<_Iter_ref_t<_CtgIt2>>, void>{};
-}
+template <class _Elem1, class _Elem2>
+struct _Lex_compare_memcmp_classify_pred<_Elem1, _Elem2, _RANGES greater> {
+    using _Pred = conditional_t<_Lex_compare_memcmp_classify_elements<_Elem1, _Elem2>, greater<int>, void>;
+};
 #endif // __cpp_lib_concepts
 
-template <class _InIt1, class _InIt2, class _Pr>
-_NODISCARD constexpr bool _Lex_compare_unchecked(
-    _InIt1 _First1, _InIt1 _Last1, _InIt2 _First2, _InIt2 _Last2, _Pr _Pred, _Lex_compare_optimize<void>) {
-    // order [_First1, _Last1) vs. [_First2, _Last2), no special optimization
-    for (; _First1 != _Last1 && _First2 != _Last2; ++_First1, (void) ++_First2) { // something to compare, do it
-        if (_DEBUG_LT_PRED(_Pred, *_First1, *_First2)) {
-            return true;
-        } else if (_Pred(*_First2, *_First1)) {
-            return false;
-        }
-    }
-
-    return _First1 == _Last1 && _First2 != _Last2;
-}
-
-template <class _CtgIt1, class _CtgIt2, class _Pr, class _Memcmp_pr>
-_NODISCARD _CONSTEXPR20 bool _Lex_compare_unchecked(
-    _CtgIt1 _First1, _CtgIt1 _Last1, _CtgIt2 _First2, _CtgIt2 _Last2, _Pr _Pred, _Lex_compare_optimize<_Memcmp_pr>) {
-    // order [_First1, _Last1) vs. [_First2, _Last2) memcmp optimization
-#if _HAS_CXX20
-    if (_STD is_constant_evaluated()) {
-        return _Lex_compare_unchecked(_First1, _Last1, _First2, _Last2, _Pred, _Lex_compare_optimize<void>{});
-    }
-#endif // _HAS_CXX20
-    (void) _Pred;
-    const auto _Num1 = static_cast<size_t>(_Last1 - _First1);
-    const auto _Num2 = static_cast<size_t>(_Last2 - _First2);
-    const int _Ans   = _Memcmp_count(_First1, _First2, (_STD min) (_Num1, _Num2));
-    return _Memcmp_pr{}(_Ans, 0) || (_Ans == 0 && _Num1 < _Num2);
-}
+template <class _It1, class _It2, class _Pred>
+using _Lex_compare_memcmp_classify =
+    conditional_t<_Iterators_are_contiguous<_It1, _It2> && !_Iterator_is_volatile<_It1> && !_Iterator_is_volatile<_It2>,
+        typename _Lex_compare_memcmp_classify_pred<_Iter_value_t<_It1>, _Iter_value_t<_It2>, _Pred>::_Pred, void>;

```

Comment:
Naming nitpick: I recommend templating on `class _Pr` here, which would be more consistent with the definition of `_Lex_compare_memcmp_classify_pred` above, and would avoid the quasi-shadowing that appears in `typename _Lex_compare_memcmp_classify_pred<MEOW, _Pred>::_Pred`.